### PR TITLE
[ENHANCEMENT][MAINTENANCE] Proper typed object for holding traversable parameters for Rule and RuleState

### PR DESCRIPTION
**Pull request by alexsherstinsky**
### Scope
* Introducing `ParameterTreeContainerNode` (as a replacement to `Parameter`).  The `ParameterTreeContainerNode` typed object is designed to depict more precisely the process of traversing the fully-qualified parameter names as "topic trees", where each node can have attribute name-value pairs (`parameters`) and can also have sub-trees (`descendants`).  The leaf nodes of this tree are the actual `stem` (non-fully-qualified) parameter names (e.g., `min`, `max`, etc.).
* Refactoring -- chancing `ConfigurationBuilder` to `ExpectationConfigurationBuilder` (based on a conversation with @anthonyburdi on 4/22/2021).

Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [ENHANCEMENT], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html).

Changes proposed in this pull request:
-
-
-

After submitting your PR, CI checks will run and @tiny-tim-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

Thank you for submitting!

### CODE REVIEW

File: great_expectations/profiler/domain_builder/simple_semantic_type_domain_builder.py [hunk 1]
```diff
@@ -40,22 +38,19 @@ def _get_domains(
...
-        type_filters: Union[str, Iterable, List[str]] = config.get("type_filters")
-        if type_filters is None:
-            # TODO: AJB 20210416 Add a test for the below comment - None = return all types
-            # None indicates no selection; all types should be returned
-            type_filters = self._type_filters
-        elif isinstance(type_filters, str):
-            type_filters = [self.SemanticDomainTypes[type_filters]]
-        elif isinstance(type_filters, Iterable):
-            type_filters = [self.SemanticDomainTypes[x] for x in type_filters]
+        semantic_types: Union[str, Iterable, List[str]] = config.get("semantic_types")
```
Comment (gilpasternak35) [inline great_expectations/profiler/domain_builder/simple_semantic_type_domain_builder.py:42]: Given that we name this variable semantic_types, does it make sense to have this variable take on a str value?

**Review by gilpasternak35 — APPROVED** [review]
LGTM!

File: great_expectations/profiler/domain_builder/simple_semantic_type_domain_builder.py (hunk 1 above)
Comment (alexsherstinsky) [reply great_expectations/profiler/domain_builder/simple_semantic_type_domain_builder.py:42]: @gilpasternak35 this came from the original prototype by James (and we use it in `DataConnector` too, perhaps in other places).  While I agree that it is not super clean, it is something we have been doing for simplicity for the users; so I solve this ambiguity by explicitly stating that its type can be a `Union` -- with that, I would feel it is acceptable.

**Review by anthonyburdi — APPROVED** [review]
LGTM!
